### PR TITLE
[test-improver] test: add enum, struct, and nullable value type edge cases for AvoidAssertAreSameWithValueTypesAnalyzer (MSTEST0038)

### PR DESCRIPTION
**Pull request by Evangelink**
## Goal and Rationale

`AvoidAssertAreSameWithValueTypesAnalyzer` (MSTEST0038) previously had only 4 tests, all using integer literals (`0`) as the representative value type. The analyzer's logic checks `IsValueType` which applies to any value type — enum, struct, and `Nullable<T>` — but none of these were exercised.

Added 3 targeted edge-case tests covering previously untested value type categories:

| New test | Value type kind | Why it matters |
|---|---|---|
| `WhenValueTypeIsEnum_Diagnostic` | Enum (`DayOfWeek`) | Enums are value types; devs often don't realize `AreSame` boxes them and always fails |
| `WhenValueTypeIsCustomStruct_Diagnostic` | User-defined `readonly struct` | Each `new MyPoint()` produces a separately boxed heap object; `AreSame` is always false at runtime |
| `WhenValueTypeIsNullableInt_Diagnostic` | `Nullable<T>` (`int?`) | `Nullable<T>` is itself a value type; `int?` variables boxed for `AreSame` are never reference-identical |

## Approach

Each test verifies both `AreSame` → `AreEqual` and `AreNotSame` → `AreNotEqual` for the given type, using the existing `VerifyCodeFixAsync` pattern.

## Trade-offs

- Tests are self-contained and small; no new dependencies.
- All 3 cases map to distinct, user-facing scenarios that the diagnostic is designed to catch.

## Reproducibility

```bash
DOTNET_INSTALL_DIR=/usr/share/dotnet ./build.sh --restore
.dotnet/dotnet run --project test/UnitTests/MSTest.Analyzers.UnitTests -f net8.0 --no-build -c Debug -- --filter "ClassName~AvoidAssertAreSameWithValueTypesAnalyzerTests"
```

## Test Status

✅ 7 tests passed (4 existing + 3 new), 0 failed.

> 🤖 **Automated content by GitHub Copilot.** Posted via a maintainer's GitHub token, so it appears under their account — the account owner did **not** write or approve this content personally. Generated by the [Test Improver](https://github.com/microsoft/testfx/actions/runs/27852600615/agentic_workflow) workflow. · 1.5K AIC · ⌖ 23.3 AIC · ⊞ 58K · [◷]( · [◷](https://github.com/search?q=repo%3Amicrosoft%2Ftestfx+%22gh-aw-workflow-id%3A+test-improver%22&type=pullrequests))
>
<details>
<summary>Add this agentic workflows to your repo</summary>

To install this agentic workflow, run

```
gh aw add githubnext/agentics/workflows/test-improver.md@main
```
</details>

### CODE REVIEW

File: test/UnitTests/MSTest.Analyzers.UnitTests/AvoidAssertAreSameWithValueTypesAnalyzerTests.cs [hunk 1]
```diff
@@ -256,4 +256,134 @@ public void TestMethod()
...
+        string code = """
+            using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+            [TestClass]
+            public class MyTestClass
+            {
+                [TestMethod]
+                public void TestMethod()
+                {
+                    int? a = 1;
```
Comment (Evangelink) [inline test/UnitTests/MSTest.Analyzers.UnitTests/AvoidAssertAreSameWithValueTypesAnalyzerTests.cs:362]: **[MODERATE] Test Completeness — null-valued nullable boundary unverified**

The test only exercises non-null `int?` values (`a = 1`, `b = 2`). Two adjacent boundaries aren't covered:

1. **Null-initialized nullable variable** — `int? a = null` should still produce a diagnostic because the *declared type* of `a` is `Nullable<int>`, which has `IsValueType == true` in Roslyn regardless of the runtime value. A test snippet like the following would confirm this:
   ```csharp
   int? a = null;
   int? b = 1;
   [|Assert.AreSame(a, b)|];   // diagnostic expected: type of a is still Nullable<int>
   ```

2. **Raw `null` literal as first argument** — `Assert.AreSame(null, b)` where `b` is `int?`: the `null` literal has no type symbol in Roslyn (`WalkDownConversion().Type` returns `null`), so `expectedType?.IsValueType` is `null` (not `true`). However, the `||` branch fires because `actualType` is `Nullable<int>`. This "only the actual side is a value type" path is exercised for integers in `UseAssertAreSameWithValueTypes_Diagnostic` (`Assert.AreSame<object>(0, "0")`), but not for `Nullable<T>` without an explicit type argument.

Neither gap is a correctness regression — the existing integer tests cover the same code paths. But a `Nullable<T>`-specific test makes the intent explicit and guards against any future `IsValueType` nuance in `Nullable`'s handling (e.g., if `WalkDownConversion()` ever strips the `Nullable` wrapper on a known-null expression).

**Review by Evangelink — COMMENTED** [review]
> [!NOTE]
> 🤖 **Automated review by GitHub Copilot.** Posted via a maintainer's GitHub token, so it appears under their account — the account owner did **not** write or approve this content personally. Generated by the [Expert Code Review workflow](https://github.com/microsoft/testfx/actions/runs/27865090001). To request a follow-up action, reply by tagging `@copilot` directly.

| # | Dimension | Verdict |
|---|-----------|---------|
| 13 | Test Completeness & Coverage | 🟡 1 MODERATE |

✅ 21/22 dimensions clean.

- [ ] Test Completeness — `WhenValueTypeIsNullableInt_Diagnostic` does not verify the null-initialized nullable case (`int? a = null`) or the untyped-null-literal-as-first-argument path (see inline comment)

---

**Full dimension breakdown**

| # | Dimension | Result |
|---|-----------|--------|
| 1 | Algorithmic Correctness | ✅ LGTM — tests correctly target enum, struct, and Nullable(T), each of which satisfies `IsValueType == true` in Roslyn; code-fix expectations are accurate |
| 2 | Threading & Concurrency | N/A — test-only change |
| 3 | Security | N/A — test-only change |
| 4 | Public API & Binary Compatibility | N/A — no production code touched |
| 5 | Performance & Allocations | N/A — test infrastructure |
| 6 | Cross-TFM Compatibility | ✅ LGTM — `readonly struct` with primary constructor in `WhenValueTypeIsCustomStruct_Diagnostic` is a C# 12 feature compiled by the Roslyn test harness; this repo's SDK targets .NET 8 (C# 12 default), so no TFM guard needed |
| 7 | Resource & IDisposable Management | N/A — no disposables |
| 8 | Defensive Coding at Boundaries | N/A — test-only change |
| 9 | Localization & Resources | N/A — no user-facing strings |
| 10 | Test Isolation | ✅ LGTM — no shared mutable state, no file-system writes, no environment mutations |
| 11 | Assertion Quality | ✅ LGTM — all tests use `VerifyCS.VerifyCodeFixAsync`, the correct harness for this project; no `BannedSymbols.txt` violations |
| 12 | Flakiness Patterns | ✅ LGTM — fully deterministic Roslyn compilation tests; no timing dependencies |
| 13 | Test Completeness & Coverage | 🟡 MODERATE — see inline comment on `WhenValueTypeIsNullableInt_Diagnostic`: null-initialized nullable variable and untyped-null-literal-as-first-arg paths are unverified |
| 14 | Data-Driven Test Coverage | ✅ LGTM — not applicable; individual diagnostic test methods are the appropriate granularity here |
| 15 | Code Structure & Simplification | ✅ LGTM — consistent structure with existing tests |
| 16 | Naming & Conventions | ✅ LGTM — `When<Condition>_Diagnostic` is the dominant pattern across the broader analyzer test suite (e.g., `ReviewAlwaysTrueAssertConditionAnalyzerTests`) |
| 17 | Documentation Accuracy | N/A — no doc comments added |
| 18 | Analyzer & Code Fix Quality | ✅ LGTM — all three tests use `VerifyCodeFixAsync` (not `VerifyAnalyzerAsync` alone), confirming both the diagnostic fires and the code fix produces correct output |
| 19 | IPC Wire Compatibility | N/A — no serialization touched |
| 20 | Build Infrastructure & Dependencies | N/A — no build files changed |
| 21 | Scope & PR Discipline | ✅ LGTM — focused, single-concern change; cleanly separated from any production logic |
| 22 | PowerShell Scripting Hygiene | N/A — no `.ps1` files |
